Push setting of default PATH down into the executor(s)

ijc opened the pull request:
Setting the default `PATH` in the `llb.State` on the client side means it
depends on the `GOOS` of the buildkit client, rather than of the environment
where it will actually execute.

Instead defer this to execution time and insert the default PATH at that point
if one is not present. Doing this in executor/oci/spec_unix.go covers both the
runc and containerd executors at once.

No special handling is needed for a Windows executor since there is no fallback
path in that case.

Fixes #604

Signed-off-by: Ian Campbell <ijc@docker.com>

ijc commented on executor/oci/spec_unix.go:
```diff
@@ -44,6 +55,10 @@ func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mou
 
+	if !hasEnv("PATH", s.Process.Env) {
+		s.Process.Env = append(s.Process.Env, "PATH="+system.DefaultPathEnv)
+	}
```
This could be achieved above with `s.Process.Env = append(s.Process.Env, meta.Env...)` instead since the containerd `GenerateSpec` also sets a default path.

I decided since there were other uses of `system.DefaultPathEnv` it was more internally consistent this way around.

tonistiigi commented:
We should try not to break the cache for this so could you add it in the `ops/exec` instead (or add in both).

ijc commented:
Ack, will move it tomorrow.

Speaking of "trying not to break things" clients after this change may not work with daemons from before it -- is that something which needs to be handled or exposed somehow?

tonistiigi commented:
> Speaking of "trying not to break things" clients after this change may not work with daemons from before it -- is that something which needs to be handled or exposed somehow?

~~Should check that we don't break the dockerfile frontend so it doesn't work with `18.06`. If needed we need can add some workaround in the frontend. Don't think we need to worry about new buildctl with old daemon.~~

Actually, should just set an apicap for this and make this compatible in client. I think it should be easy. If there are issues we can do the strikethrough option.

ijc commented:
I think I either don't understand the apicap usage or this would require a breaking change to the API. I expected to have to do something like:
```diff
func NewState(o Output) State {
        s := State{
                out: o,
                ctx: context.Background(),
        }
        s = dir("/")(s)
-       s = addEnv("PATH", system.DefaultPathEnv)(s)
+       if !HasCap(AutomaticFallbackPath) {
+               s = addEnv("PATH", system.DefaultPathEnv)(s)
+       }
        s = s.ensurePlatform()
        return s
}
```
but there is global `HasCap` type function I can see. The only thing I can find is `c.caps.Supports`, but that is only a gateway thing (not a control API one) and in any case we don't have any sort of suitable `c` in our hands to use here and adding one would change the client API for user code.

I'm not sure if I'm supposed to do something with constraints, or the `s.opts` but even if so I can't see anywhere on the client side where it would be possible to correlate the server caps with that and to apply the necessary fallback (nowhere relevant, like the `Marshal` method, has a handle on any sort of server cap). I don't think this case is covered by #465, or at least I'm not seeing it.

I suppose `grpcClient.Solve` could munge `Definition` somehow -- but that would involve unmarshalling and remarshalling it which doesn't seem at all desirable.

Am I missing something or do we need to go with the struckout version?

tonistiigi commented:
@ijc Indeed didn't remember that `InfoRequest` was not implemented https://github.com/moby/buildkit/blob/master/api/services/control/control.proto#L24 . Created https://github.com/moby/buildkit/pull/641 that at least allows to check caps when using the gateway API. In `llb/exec/marshal()` you can check `c.Caps`

ijc commented:
Updated to build upon #641, so with client backwards compat added.

I moved the fallback to `solver/llbsolver/ops/exec.go` as suggested. Is that the right place (or at least not an unchangeably wrong place) WRT future plans for remote workers with different platforms?

tonistiigi commented on client/llb/exec.go:
```diff
@@ -142,6 +143,11 @@ func (e *ExecOp) Marshal(c *Constraints) (digest.Digest, []byte, *pb.OpMetadata,
 		}
 	}
+	if err := c.Caps.Supports(pb.CapExecMetaSetsDefaultPath); err != nil {
+		e.meta.Env.SetDefault("PATH", system.DefaultPathEnv)
```
This shouldn't work because `meta.Env` is not a pointer.

ijc replied:
Urk, I think I need to redouble my efforts to construct a test case then!

tonistiigi commented on client/llb/exec.go:
```diff
@@ -142,6 +143,11 @@ func (e *ExecOp) Marshal(c *Constraints) (digest.Digest, []byte, *pb.OpMetadata,
 			e.meta.Env = e.meta.Env.AddOrReplace("SSH_AUTH_SOCK", e.ssh[0].Target)
 		}
 	}
+	if err := c.Caps.Supports(pb.CapExecMetaSetsDefaultPath); err != nil {
```
What behavior makes more sense here if `c.Caps` is nil (eg. offline on a direct client with no filtering set). I think we could assume up-to-date daemon then.

ijc replied:
#641 requires the user to have known (and remembered) to pass `llb.WithCaps(caps)` to `Marshal` here. If they do not (i.e. all preexisting code) then `c.Caps` will be `nil` and they would be expecting the legacy behaviour -- i.e. CapExecMetaSetsDefaultPath not present so they get a default PATH set at this level.

There's a good argument that if `WithCaps` isn't used then `Marshall` should inject a set of caps which corresponds to the state at the point `WithCaps` was added, but that also wouldn't include `CapExecMetaSetsDefaultPath` so while it is a good idea IMHO it's a bit moot WRT this PR.

tonistiigi replied:
> then c.Caps will be nil and they would be expecting the legacy behaviour -- i.e. CapExecMetaSetsDefaultPath not present so they get a default PATH set at this level.

I think they shouldn't get the default path. If they explicitly say that they have a set of caps that doesn't contain this we can skip it. This is more in line with the other features that just add the caps they use for validation instead of expecting that they are not supported if caps is not set. So I think this can just be wrapped in `if c.Caps != nil`.

ijc replied:
OK. I also interpreted this as implying I should drop `Allow `apicaps.CapSet.Supports` to be called on a nil CapSet.` from this PR, so I will do that too.
```diff
commit 0908e733f98eac440ac8f472483313f275ef016a
Author: Ian Campbell <ijc@docker.com>
Date:   Tue Sep 25 10:46:09 2018 +0100

    Allow `apicaps.CapSet.Supports` to be called on a nil CapSet.
    
    Signed-off-by: Ian Campbell <ijc@docker.com>

diff --git a/util/apicaps/caps.go b/util/apicaps/caps.go
index 9a661b50..9ead25ca 100644
--- a/util/apicaps/caps.go
+++ b/util/apicaps/caps.go
@@ -106,6 +106,9 @@ type CapSet struct {
 // Supports returns an error if capability is not supported
 func (s *CapSet) Supports(id CapID) error {
        err := &CapError{ID: id}
+       if s == nil || s.list == nil {
+               return err
+       }
        c, ok := s.list.m[id]
        if !ok {
                return errors.WithStack(err)
```

ijc commented:
I've added a test _and_ made it pass!
